lib/src/supertokens-http-client.dart

```diff
@@ -248,22 +278,43 @@ class Client extends http.BaseClient {
     }
   }
 
-  static Future<void> saveTokensFromHeaders(
-      http.StreamedResponse response) async {
-    String? frontTokenFromResponse = response.headers[frontTokenHeaderKey];
-    if (frontTokenFromResponse != null) {
-      await FrontToken.setItem(frontTokenFromResponse);
+  static Future clearTokensIfRequired() async {
+    LocalSessionState preRequestLocalSessionState =
+        await SuperTokensUtils.getLocalSessionState();
+    if (preRequestLocalSessionState.status ==
+        LocalSessionStateStatus.NOT_EXISTS) {
+      await AntiCSRF.removeToken();
+      await FrontToken.removeToken();
     }
+  }
 
-    String? antiCSRFFromResponse = response.headers[antiCSRFHeaderKey];
-    if (antiCSRFFromResponse != null) {
-      LocalSessionState localSessionState =
-          await SuperTokensUtils.getLocalSessionState();
-      await AntiCSRF.setToken(
-        antiCSRFFromResponse,
-        localSessionState.lastAccessTokenUpdate,
-      );
-    }
+  static Map<String, dynamic> _generateMapFromCookie(String cookie) {
+    if (cookie.isEmpty) return {};
+    List<String> entries = cookie.split(";");
+    Map<String, dynamic> map = {};
+    entries.forEach((element) {
+      List<String> keyValuePair = element.split("=");
+      if (keyValuePair.length == 2) {
+        map[keyValuePair[0]] = keyValuePair[1];
+      }
+    });
+    return map;
+  }
+
+  static String _generateCookieHeader(String? oldCookie, String? newCookie) {
+    Map<String, dynamic> existingCookieHeaderMap =
+        _generateMapFromCookie(oldCookie ?? "");
+    Map<String, dynamic> newCookiesToAddMap =
+        _generateMapFromCookie(newCookie ?? "");
+    String finalCookieHeader = "";
+    existingCookieHeaderMap.keys.forEach((element) {
+      if (newCookiesToAddMap.containsKey(element)) {
```

Comment:
Cookies inside newCookiesToAddMap that are not inside existingCookieHeaderMap are never added. I dont like this approach in general, follow these steps

1. Parse both sets of cookies
2. filter existing cookies to not include any that are present in the new cookies
3. Merge the two sets of cookies
4. Create a string